static/app/utils/useProjects.tsx

```diff
@@ -1,13 +1,315 @@
+  initiallyLoaded: boolean;
+  /**
+   * This is state for when fetching data from API
+   */
+  fetching: boolean;
+  /**
+   * The error that occurred if fetching failed
+   */
+  fetchError: null | RequestError;
+  /**
+   * Indicates that Project results (from API) are paginated and there are more
+   * projects that are not in the initial response
+   */
+  hasMore: null | boolean;
+  /**
+   * The last query we searched. Used to validate the cursor
+   */
+  lastSearch: null | string;
+  /**
+   * Pagination
+   */
+  nextCursor?: null | string;
+};
+
+type Result = {
+  /**
+   * The loaded projects list
+   */
+  projects: Project[];
+  /**
+   * When loading speciific slugs, placeholder objects will be returned
+   */
+  placeholders: ProjectPlaceholder[];
+  /**
+   * This is an action provided to consumers for them to update the current
+   * projects result set using a simple search query.
+   *
+   * Will always add new options into the store.
+   */
+  onSearch: (searchTerm: string) => Promise<void>;
+} & Pick<State, 'fetching' | 'hasMore' | 'fetchError' | 'initiallyLoaded'>;
+
+type Options = {
+  /**
+   * Specify an orgId, overriding the organization in the current context
+   */
+  orgId?: string;
+  /**
+   * List of slugs to look for summaries for, this can be from `props.projects`,
+   * otherwise fetch from API
+   */
+  slugs?: string[];
+  /**
+   * Number of projects to return when not using `props.slugs`
+   */
+  limit?: number;
+};
+
+type FetchProjectsOptions = {
+  slugs?: string[];
+  limit?: Options['limit'];
+  cursor?: State['nextCursor'];
+  search?: State['lastSearch'];
+  lastSearch?: State['lastSearch'];
+};
+
+/**
+ * Helper function to actually load projects
+ */
+async function fetchProjects(
+  api: Client,
+  orgId: string,
+  {slugs, search, limit, lastSearch, cursor}: FetchProjectsOptions = {}
+) {
+  const query: {
+    collapse: string[];
+    query?: string;
+    cursor?: typeof cursor;
+    per_page?: number;
+    all_projects?: number;
+  } = {
+    // Never return latestDeploys project property from api
+    collapse: ['latestDeploys'],
+  };
+
+  if (slugs !== undefined && slugs.length > 0) {
+    query.query = slugs.map(slug => `slug:${slug}`).join(' ');
+  }
+
+  if (search) {
+    query.query = `${query.query ?? ''}${search}`.trim();
+  }
+
+  const prevSearchMatches = (!lastSearch && !search) || lastSearch === search;
+
+  if (prevSearchMatches && cursor) {
+    query.cursor = cursor;
+  }
+
+  if (limit !== undefined) {
+    query.per_page = limit;
+  }
+
+  let hasMore: null | boolean = false;
+  let nextCursor: null | string = null;
+  const [data, , resp] = await api.requestPromise(`/organizations/${orgId}/projects/`, {
+    includeAllArgs: true,
+    query,
+  });
+
+  const pageLinks = resp?.getResponseHeader('Link');
+  if (pageLinks) {
+    const paginationObject = parseLinkHeader(pageLinks);
+    hasMore = paginationObject?.next?.results || paginationObject?.previous?.results;
```

Comment:
Should you be looking at previous results? Do we need to go backwards with pagination?

Reply:
Ah good question. useTeams does this too 🤔

Reply:
Will try and fix this up separately I think, and leave this consistent w/ useTeams for now.